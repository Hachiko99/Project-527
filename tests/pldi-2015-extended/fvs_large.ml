type nat =
  | O
  | S of nat

type list =
  | Nil
  | Cons of nat * list

type exp =
  | Unit
  | BVar of nat
  | FVar of nat
  | Lam of nat * exp
  | App of exp * exp
  | Pair of exp * exp
  | Fst of exp
  | Snd of exp
  | Inl of exp
  | Inr of exp
  | Const of nat
  | Add of exp * exp
  | Sub of exp * exp
  | Mult of exp * exp
  | Div of exp * exp

let rec append (l1:list) (l2:list) : list =
  match l1 with
  | Nil -> l2
  | Cons (x, l1p) -> Cons (x, append l1p l2)
;;

let fvs_large : exp -> list |>
  { Unit => []
  | FVar (0) => [0]
  | FVar (1) => [1]
  | FVar (2) => [2]
  | BVar (0) => []
  | Lam (0, Unit) => []
  | Lam (0, FVar (1)) => [1]
  | App (Unit, Unit) => []
  | App (FVar (0), Unit) => [0]
  | App (Unit, FVar (1)) => [1]
  | Fst (Unit) => []
  | Fst (FVar (1)) => [1]
  | Snd (Unit) => []
  | Snd (FVar (1)) => [1]
  | Pair (Unit, Unit) => []
  | Pair (FVar (0), Unit) => [0]
  | Pair (Unit, FVar (1)) => [1]
  | Pair (FVar (0), FVar (1)) => [0; 1]
  | Inl (Unit) => []
  | Inl (FVar (1)) => [1]
  | Inr (Unit) => []
  | Inr (FVar (1)) => [1]
  | Const (0) => []
  | Add (FVar (0), Unit) => [0]
  | Add (Unit, FVar (1)) => [1]
  | Sub (FVar (0), Unit) => [0]
  | Sub (Unit, FVar (1)) => [1]
  | Mult (FVar (0), Unit) => [0]
  | Mult (Unit, FVar (1)) => [1]
  | Div (FVar (0), Unit) => [0]
  | Div (Unit, FVar (1)) => [1]
  } = ?
