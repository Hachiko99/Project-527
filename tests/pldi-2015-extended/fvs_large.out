(* Generated Tue Jul 21 06:18:06 2015 *)
let fvs_large : exp -> list =
  let rec f1 (e1:exp) : list =
    match e1 with
      | Unit -> Nil
      | BVar (n1) -> Nil
      | FVar (n1) -> Cons (n1, Nil)
      | Lam (n1, e2) -> f1 e2
      | App (e2, e3) -> (match f1 e3 with
                           | Nil -> f1 e2
                           | Cons (n1, l1) -> Cons (S (O), Nil))
      | Pair (e2, e3) -> (match f1 e2 with
                            | Nil -> f1 e3
                            | Cons (n1, l1) -> Cons (O, f1 e3))
      | Fst (e2) -> f1 e2
      | Snd (e2) -> f1 e2
      | Inl (e2) -> f1 e2
      | Inr (e2) -> f1 e2
      | Const (n1) -> Nil
      | Add (e2, e3) -> (match f1 e3 with
                           | Nil -> Cons (O, Nil)
                           | Cons (n1, l1) -> Cons (S (O), Nil))
      | Sub (e2, e3) -> (match f1 e3 with
                           | Nil -> Cons (O, Nil)
                           | Cons (n1, l1) -> Cons (S (O), Nil))
      | Mult (e2, e3) -> (match f1 e3 with
                            | Nil -> Cons (O, Nil)
                            | Cons (n1, l1) -> Cons (S (O), Nil))
      | Div (e2, e3) -> (match f1 e3 with
                           | Nil -> Cons (O, Nil)
                           | Cons (n1, l1) -> Cons (S (O), Nil))
  in
    f1
;;
